src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs

```diff
@@ -120,19 +123,24 @@ internal void AttachTokens()
                         ExpirationTokenRegistrations.Add(registration);
                     }
                 }
+                _lock.Release();
             }
         }
 
         private static void ExpirationTokensExpired(object obj)
         {
-            var entry = (CacheEntry)obj;
-            entry.SetExpired(EvictionReason.TokenExpired);
-            entry._notifyCacheOfExpiration(entry);
+            Task.Factory.StartNew(() =>
+            {
+                var entry = (CacheEntry)obj;
+                entry.SetExpired(EvictionReason.TokenExpired);
+                entry._notifyCacheOfExpiration(entry);
+            }, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
```

Comment:
Flow the state to avoid the closure: https://msdn.microsoft.com/en-us/library/dd321456(v=vs.110).aspx